### PR TITLE
Improve AdditionalAssemblies coverage

### LINKED ISSUE #31081: BWA cient project guidance for AdditionalAssemblies
### Description

In versioned content, clarify how to use the `.Client` project's `_Imports` file in the assembly argument to `AdditionalAssemblies`.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#route-to-components-from-multiple-assemblies

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/fundamentals/routing.md

### Document ID

b6cb2574-6326-a40d-0b6c-d1661328e828

### Article author

guardrex

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #31081
Addresses #28161

We'll need coverage on **static** versus **interactive** routers.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/components/render-modes.md](https://github.com/dotnet/AspNetCore.Docs/blob/1e3b7255515274a06fb9abb221f1ea95ac2d9598/aspnetcore/blazor/components/render-modes.md) | [ASP.NET Core Blazor render modes](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?branch=pr-en-us-31131) |
| [aspnetcore/blazor/fundamentals/routing.md](https://github.com/dotnet/AspNetCore.Docs/blob/1e3b7255515274a06fb9abb221f1ea95ac2d9598/aspnetcore/blazor/fundamentals/routing.md) | [ASP.NET Core Blazor routing and navigation](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?branch=pr-en-us-31131) |
| [aspnetcore/migration/70-80.md](https://github.com/dotnet/AspNetCore.Docs/blob/1e3b7255515274a06fb9abb221f1ea95ac2d9598/aspnetcore/migration/70-80.md) | [[Visual Studio Code](#tab/visual-studio-code)](https://review.learn.microsoft.com/en-us/aspnet/core/migration/70-80?branch=pr-en-us-31131) |

### CODE REVIEW

File: aspnetcore/blazor/fundamentals/routing.md [hunk 1]
```diff
@@ -135,30 +135,37 @@ Arbitrary items are supported as content of the `<NotFound>` tags, such as other
 
 ## Route to components from multiple assemblies
```
Comment (danroth27) [inline aspnetcore/blazor/fundamentals/routing.md:136]: Can we cover both `Router.AdditionalAssemblies` and `MapRazorComponents<TComponent>().AddAdditionalAssemblies()` in this section? The former is used when the Blazor `Router` is handling page navigations interactively. The latter is used when routing requests to components on the server using endpoint routing.

Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md:136]: I moved the coverage over to this section and left a cross-link to this content in the *Render Modes* article.

File: aspnetcore/blazor/fundamentals/routing.md [hunk 2]
```diff
@@ -135,30 +135,43 @@ Arbitrary items are supported as content of the `<NotFound>` tags, such as other
...
+    AdditionalAssemblies="new[] { typeof(BlazorSample.Client._Imports).Assembly }">
+    ...
 </Router>
 ```
 
+A Blazor Web App is required to indicate the `.Client` project's assembly to <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> if the Blazor router is interactive, either on the server or on the client.
+
+If the the Blazor router isn't an interactive router, routable components of the `.Client` project are discovered without specifying the `.Client` project's assembly to <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies>.
+
+To discover components from additional assemblies for static server-side rendering, identify the additional assemblies with the <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A> method chained to <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents%2A>.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/routing.md:166]: This first paragraph makes it sound to me like, if the router is not interactive, you don't have to do anything to have the `.Client` project's routable components get discovered. Then the second paragraph indicates that you _do_ have to do something, and that's registering the additional assemblies using `AddAdditionalAssemblies`. I wonder if there's a way to unify those somehow. Or if you think it's clear enough already, no need to try to reword it!

I've also just written up an example below of how I would explain the concept of "additional assemblies" as it pertains to various app configurations. Feel free to ignore it completely if you like what you have, but I'm putting it here in case you want to borrow any ideas from it:

---

If the router is interactive, any additional assemblies containing routable components must be listed in <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies>.

To have routable components from additional assemblies get discovered for static server rendering, those assemblies must be listed in <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A>.

For example, in a Blazor Web app, an interactive render mode can be specified on the `Routes` component, which in turn renders the `Router` component. If the `Routes` component is defined in the server project, the <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> parameter should include the `.Client` assembly. This allows the router to work correctly even when rendered interactively.

However, in a Blazor Web app with global WebAssembly interactivity enabled, the `Routes` component might be defined in the `.Client` project. In this case, there aren't assemblies external to the one defining the `Routes` component that contain routable components, so it's not necessary to specify a value for <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies>.

In both cases, though, it's necessary to include routable components in the `.Client` assembly for static server rendering using <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A>. This is because the initial, static server-side render needs to be able to identify components in the `.Client` assembly, even if the router later becomes interactive.

**Review by MackinnonBuck — APPROVED** [review]

File: aspnetcore/blazor/fundamentals/routing.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md:166]: That's ***much*** better. I'll place your text for the coverage.

WRT "static" versus "interactive" routers: Can you describe them for me?

File: aspnetcore/blazor/components/render-modes.md [hunk 3]
```diff
@@ -532,16 +532,9 @@ The following component results in a runtime error when the component is rendere
...
 
-Configure additional assemblies to discover routable Razor components for Static Server rendering using the `AddAdditionalAssemblies` method chained to <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents%2A>.
-
-The following example includes the assembly of the `DifferentAssemblyCounter` component:
-
-```csharp
-app.MapRazorComponents<App>()
-    .AddAdditionalAssemblies(typeof(DifferentAssemblyCounter).Assembly);
-```
+Additional assemblies must be identified to the Blazor framework to discover routable Razor components in referenced projects. For more information, see <xref:blazor/fundamentals/routing#route-to-components-from-multiple-assemblies>.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/components/render-modes.md:537]:
```suggestion
Additional assemblies must be supplied to the Blazor framework to discover routable Razor components in referenced projects. For more information, see <xref:blazor/fundamentals/routing#route-to-components-from-multiple-assemblies>.
```
(or maybe "identified by" or "identifiable by" would work as well)

File: aspnetcore/blazor/fundamentals/routing.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md]: Nice ... I'll roll this to the new section at the top of the article first thing Wednesday morning.

File: aspnetcore/blazor/fundamentals/routing.md [hunk 4]
```diff
@@ -19,6 +19,18 @@ This article explains how to manage Blazor app request routing and how to use th
...
+
+## Static versus interactive routing
+
+*This section applies to Blazor Web Apps.*
+
+If [prerendering isn't disabled](xref:blazor/components/render-modes#prerendering), the Blazor router (`Router` component, `<Router>` in `Routes.razor`) performs static routing to components during static server-side rendering (static SSR). This type of routing is called *static routing*, and the Blazor router is called a *static router* when static routing is taking place.
+
+An interactive render mode can be assigned to the `Routes` component that makes the Blazor router become interactive after static SSR and static routing on the server. For example, `<Routes @rendermode="InteractiveServer" />` assigns the Interactive Server render mode to the `Routes` component. The `Router` component inherits the Interactive Server render mode from the `Routes` component. The router becomes *interactive* after static routing on the server. This type of routing is called *interactive routing*, and the Blazor router is called an *interactive router*. An interactive router can route requests either on the server, *interactive server-side routing*, or on the client, *interactive client-side routing*.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/routing.md:30]: > after static SSR and static routing on the server

We might be able to just shorten this to "after static SSR" because static server-side routing is just one of things that happens during static SSR, and isn't a separate step. But I could go either way on this and I can see the benefit of explicitly calling out that static server side routing happens before interactivity starts.

> The router becomes *interactive* after static routing on the server

I think "The router becomes *interactive* after static SSR" might be more accurate, because the routing step happens before static SSR completes. It could be possible to interpret the existing wording as meaning "The router becomes interactive _immediately_ after static routing on the server", which wouldn't be accurate.

> An interactive router can route requests either on the server, *interactive server-side routing*, or on the client, *interactive client-side routing*

Do we use this terminology elsewhere? I've been personally thinking of all interactive routing as "client side routing" regardless of whether the router uses Server or WebAssembly interactivity.

File: aspnetcore/blazor/fundamentals/routing.md [hunk 5]
```diff
@@ -135,30 +147,69 @@ Arbitrary items are supported as content of the `<NotFound>` tags, such as other
...
 
-Use the <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> parameter to specify additional assemblies for the <xref:Microsoft.AspNetCore.Components.Routing.Router> component to consider when searching for routable components. Additional assemblies are scanned in addition to the assembly specified to `AppAssembly`. In the following example, `Component1` is a routable component defined in a referenced [component class library](xref:blazor/components/class-libraries). The following <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> example results in routing support for `Component1`.
+:::moniker range=">= aspnetcore-8.0"
 
-:::moniker range=">= aspnetcore-6.0"
+*This section applies to Blazor Web Apps.*
+
+Use the <xref:Microsoft.AspNetCore.Components.Routing.Router> component's <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> parameter and the endpoint convention builder <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A> to discover routable components in additional assemblies. The following subsections explain when and how to use the API.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/routing.md:154]:
```suggestion
Use the <xref:Microsoft.AspNetCore.Components.Routing.Router> component's <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> parameter and the endpoint convention builder <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A> to discover routable components in additional assemblies. The following subsections explain when and how to use each API.
```

File: aspnetcore/blazor/fundamentals/routing.md [hunk 6]
```diff
@@ -135,30 +147,69 @@ Arbitrary items are supported as content of the `<NotFound>` tags, such as other
...
+
+### Static routing
+
+To discover routable components from additional assemblies for static server-side rendering (static SSR), even if the router later becomes interactive for interactive rendering, the assemblies must be identified to the Blazor framework. Call the <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A> method with the additional assemblies chained to <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents%2A> in the server project's `Program` file.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/routing.md:158]: > must be identified to the Blazor framework

Similar to my first suggestion, maybe "provided to", "identified by", or "identifiable by" might be clearer. Just a nit though so feel free to ignore if you disagree!

File: aspnetcore/blazor/fundamentals/routing.md [hunk 7]
```diff
@@ -135,30 +147,69 @@ Arbitrary items are supported as content of the `<NotFound>` tags, such as other
...
-    AppAssembly="@typeof(App).Assembly"
-    AdditionalAssemblies="new[] { typeof(Component1).Assembly }">
-    @* ... Router component elements ... *@
+    AppAssembly="..."
+    AdditionalAssemblies="new[] { typeof(BlazorSample.Client._Imports).Assembly }">
+    ...
 </Router>
 ```
 
+Additional assemblies are scanned in addition to the assembly specified to <xref:Microsoft.AspNetCore.Components.Routing.Router.AppAssembly%2A>.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/routing.md:186]: 👍

File: aspnetcore/blazor/components/render-modes.md (hunk 3 above)
Comment (guardrex) [reply aspnetcore/blazor/components/render-modes.md:537]: Let's try "disclosed" here.

File: aspnetcore/blazor/fundamentals/routing.md (hunk 6 above)
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/routing.md:158]:
```suggestion
To discover routable components from additional assemblies for static server-side rendering (static SSR), even if the router later becomes interactive for interactive rendering, the assemblies must be disclosed to the Blazor framework. Call the <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointConventionBuilderExtensions.AddAdditionalAssemblies%2A> method with the additional assemblies chained to <xref:Microsoft.AspNetCore.Builder.RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents%2A> in the server project's `Program` file.
```

File: aspnetcore/blazor/fundamentals/routing.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md:30]: WRT both terms, we don't use "interactive server-side routing/router" anywhere. We used to use "interactive client routing"/"interactive client-side router" in one sentence in the enhanced nav content, ***but*** we 🔪 that out on that other PR that killed that one confusing paragraph. Now, we don't use any of these terms.

However, it seems unbalanced not to have the terminology available (if it's a thing) because if the app isn't running on the client at all, then there can't be interactive client-side routing, right? Which of the following paradigms matches what the framework does for mixed/Auto render mode apps ...

### There's no such thing as interactive server-side routing beyond static routing

* Static routing/router on the server 👉 Interactive client-side routing/router on the client for WASM-rendered components
* Static routing/router on the server

### Interactive server-side routing occurs after static routing

* Static routing/router on the server 👉 Interactive client-side routing/router on the client for WASM-rendered components
* Static routing/router on the server 👉 Interactive server-side routing/router on the server for server-rendered components

Comment (MackinnonBuck) [reply aspnetcore/blazor/fundamentals/routing.md:30]: The second option you wrote definitely looks more accurate to me, so I'd be happy with going with that.

I've personally been using the term "client-side routing" to mean "any routing that works separately from endpoint routing". For that reason, I would view interactive routing via server interactivity still as a form of "client-side routing". Internal navigations get intercepted, and page content gets generated dynamically by the client (browser) rather than being provided by the server. But maybe others would disagree! @danroth27, do you have an opinion on this?

File: aspnetcore/blazor/fundamentals/routing.md [hunk 8]
```diff
@@ -19,6 +19,18 @@ This article explains how to manage Blazor app request routing and how to use th
 > [!IMPORTANT]
 > Code examples throughout this article show methods called on `Navigation`, which is an injected <xref:Microsoft.AspNetCore.Components.NavigationManager> in classes and components.
 
+:::moniker range=">= aspnetcore-8.0"
+
+## Static versus interactive routing
+
+*This section applies to Blazor Web Apps.*
+
+If [prerendering isn't disabled](xref:blazor/components/render-modes#prerendering), the Blazor router (`Router` component, `<Router>` in `Routes.razor`) performs static routing to components during static server-side rendering (static SSR). This type of routing is called *static routing*, and the Blazor router is called a *static router* when static routing is taking place.
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/routing.md:28]:
```suggestion
If [prerendering isn't disabled](xref:blazor/components/render-modes#prerendering), the Blazor router (`Router` component, `<Router>` in `Routes.razor`) performs static routing to components during static server-side rendering (static SSR). This type of routing is called *static routing*.
```

File: aspnetcore/blazor/fundamentals/routing.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md:30]: @MackinnonBuck ... I don't think DR will be able to respond due to prior commitments. However, we're close to working this out.

I simplified the text above, distilling it down to just static and interactive routing without getting into the weeds on server/client location. See how that looks, and I'll take another look to see if whatever we say about interactive ***client-side*** routing makes sense based on these remarks that we're working on now.

Comment (guardrex) [reply aspnetcore/blazor/fundamentals/routing.md:30]: I'm going to go ahead with this. If there's a concern about the final commit, I'll set up a patch PR.
